core/domain/exp_domain.py

```diff
@@ -2968,43 +2968,62 @@ def delete_state(self, state_name):
         del self.states[state_name]
 
 
-    def get_trainable_states_dict(self, old_states, change_list):
+    def get_state_names_mapping(self, change_list):
+        """Creates a mapping from the current state names of the exploration
+        to their corresponding state names of the older version.
+
+        Args:
+            change_list: list(dict). A list of changes introduced in this
+                commit.
+
+        Returns:
+            dict. The state_names_mapping dict.
+        """
+        state_names_mapping = {}
+        renamed_cmds = []
+        for change_dict in change_list:
```

Comment:
There's a simpler algorithm to do this that doesn't involve keeping track of a separate list. Start with a dict mapping each state name to itself (this is an old-name --> new-name mapping). Iterate _backwards_ along the change list. When you come to a rename cmd, update that state name's _key_ (old name) in the dict. At the end, flip the dict so that you get a new-name --> old-name dict. This is O(N) instead of O(N^2).